src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,256 @@
+
+                            // Safe because the task is completed.
+                            return funcTask.Result;
+                        }
+                        else if (completedTask.Status == TaskStatus.Faulted)
+                        {
+                            tasks.Remove(completedTask);
+                            tasks.Remove(parallelAwait);
+
+                            if (IsRetryableException(completedTask.Exception))
+                            {
+                                lastException = completedTask.Exception;
+                                client.FailedAttempts++;
+                                TimeSpan backoffInterval = BackoffIntervalConstants.MinBackoffInterval.CalculateBackoffInterval(BackoffIntervalConstants.MaxBackoffInterval, client.FailedAttempts);
+                                client.BackoffEndTime = DateTimeOffset.UtcNow.Add(backoffInterval);
+                            }
+                            else
+                            {
+                                completedTask.Exception.Handle(e =>
+                                {
+                                    // propagate the original exception.
+                                    throw e;
+                                });
+                            }
+                        }
+                        else if (completedTask.Status == TaskStatus.Canceled)
+                        {
+                            // Throw expected OperationCanceledException
+                            completedTask.GetAwaiter().GetResult();
+                        }
+                    }
+                    else
+                    {
+                        tasks.Remove(parallelAwait);
+                    }
+                }
+            }
+
+            throw lastException;
+        }
+
+        private bool IsRetryableException(AggregateException ex)
+        {
+            int statusCode = 0;
+
+            IReadOnlyCollection<Exception> innerExceptions = ex.Flatten().InnerExceptions;
+
+            if (innerExceptions.Count > 0 && innerExceptions.All(ex => ex is RequestFailedException))
+            {
+                statusCode = (innerExceptions.Last() as RequestFailedException).Status;
+            }
+
+            switch (statusCode)
+            {
+                case (int)HttpStatusCode.RequestTimeout:       // 408
+                case HttpStatusCodeRequestThrottled:           // 429
+                case (int)HttpStatusCode.InternalServerError:  // 500
+                case (int)HttpStatusCode.BadGateway:           // 502
+                case (int)HttpStatusCode.ServiceUnavailable:   // 503
+                case (int)HttpStatusCode.GatewayTimeout:       // 504
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        private IEnumerable<ConfigurationClientState> GetPrioritizedConfigurationClientList()
```

Comment:
I'm wondering if it's possible the BackoffEndTime of clients are updated while this function is running.

Reply:
Jimmy pointed out that we don't allow simultaneous network operations. (reference)[[https://github.com/Azure/AppConfiguration-DotnetProvider/blob/0ac6b93e7abbb31fbb46090baf244dcc91384447/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs#L42](https://github.com/Azure/AppConfiguration-DotnetProvider/blob/0ac6b93e7abbb31fbb46090baf244dcc91384447/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs#L42)]

That means, we won't have multiple threads fiddling with this class, and it isn't possible for the BackoffTime to be updated while this function is running.